### PR TITLE
[improve][admin,broker] Add option to unloadNamespaceBundle with bundle Affinity broker url

### PR DESCRIPTION
**Pull request by vineeth1995**
### Motivation

To provide an option for admin to unload namespace bundle with broker url to which the bundle has to be assigned to after the unload.

### Modifications

1. Add an option to provide brokerWebServiceAddress (Ex: https://broker1.com:4443) while unloadingNamespaceBundle in the admin CLI and make corresponding changes in the pulsar-client-admin-api to support this.

2. Changes in pulsar-broker unloadNamespaceBundle API to accommodate  broker url while unloading. 

3. Add functionality in LoadManger implementation to select affinity broker for a bundle while trying to assign unloaded bundle to broker.

### Verifying this change

This change added tests and can be verified as follows:

  - Added end to end test to make sure the bundle assignment happens as per the expectation

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [x] The REST endpoints
- [x] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [ ] `doc-not-needed` 
- [x] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Comment by codelipenghui** [comment]
@vineeth1995 Thanks for the contribution. 
The new feature should start with a proposal https://github.com/apache/pulsar/blob/master/wiki/proposals/PIP.md
So that we can understand the motivation for the API change.

**Review by rdhabalia — APPROVED** [review]
LGTM

**Review by codelipenghui — CHANGES_REQUESTED** [review]
A proposal is required since the PR is changed the Admin API and introduced a new feature.

**Comment by rdhabalia** [comment]
@codelipenghui I don't think this will require PIP as it's part of adding a flag in admin API. There are a large number of PRs which must have gone through PIP with similar changes but merge without any explanation. let me give you few examples
https://github.com/apache/pulsar/pull/16167
https://github.com/apache/pulsar/pull/14680
https://github.com/apache/pulsar/pull/12136
https://github.com/apache/pulsar/pull/13938

also PR which breaks backward compatibility https://github.com/apache/pulsar/pull/10601 was merged without any concerns. 
I have many such examples where I see practice by the specific community to discourage people from their contribution and delay their efforts. I would strongly discourage such destructive practices. 
There are such bad practices going on in the community and I have many examples. Please let us know if you have any concerns with the PR by adding an option flag if not then you should not BLOCK the PR unnecessarily as it will start -ve and destructive practice in the community and we didn't open source Pulsar for such things.

**Comment by rdhabalia** [comment]
@codelipenghui I know there were multiple genuine PRs were blocked in past and they were sitting unmerged for a long time for multiple releases. I can clearly see the intention but we should avoid such destructive practice instead we should spend time in constructive efforts as not every contributor is vendor but users/companies are trying to contribute for their business usecases and such practice discourages companies to use Pulsar.

**Comment by rdhabalia** [comment]
@codelipenghui can you please unblock the PR?

**Comment by rdhabalia** [comment]
@codelipenghui ping

**Comment by rdhabalia** [comment]
this PR is blocked for last 3 days and I haven't heard anything from @codelipenghui . can we unblock the PR or dismiss the review? we can't keep following up on the same thing again and again.

**Comment by codelipenghui** [comment]
> @codelipenghui I don't think this will require PIP as it's part of adding a flag in admin API. There are a large number of PRs which must have gone through PIP with similar changes but merge without any explanation. let me give you few examples
https://github.com/apache/pulsar/pull/16167
https://github.com/apache/pulsar/pull/14680
https://github.com/apache/pulsar/pull/12136
https://github.com/apache/pulsar/pull/13938

> also PR which breaks backward compatibility https://github.com/apache/pulsar/pull/10601 was merged without any concerns. 
I have many such examples where I see practice by the specific community to discourage people from their contribution and delay their efforts. I would strongly discourage such destructive practices. 
There are such bad practices going on in the community and I have many examples. Please let us know if you have any concerns with the PR by adding an option flag if not then you should not BLOCK the PR unnecessarily as it will start -ve and destructive practice in the community and we didn't open source Pulsar for such things.

Sorry for the late response.

Yes, this is why I think it's better to have a proposal to avoid we will not introduce any API changes without a proposal in the future. I agree there are many changes to admin APIs without proposals/discussions. But we are on the wrong way, right? More and more people are starting to use Pulsar. We should be more careful about the API changes than ever. 

I want to say there have been many CLI, and admin API changes without proposals before. If you just don't think we should not start with this PR(proposal required). It's ok for me. I think I should start a discussion under the mailing list first, not block this PR under an uncleared rule. I see all the PRs you provided are from StreamNative's engineers.  Please don't think I mean to facilitate StreamNative engineers. 

Here is an example of a proposal that is not required for new metrics. 

https://github.com/apache/pulsar/issues/18319
https://github.com/apache/pulsar/issues/18560

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java:1452]: StringUtils.isBlank

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java:1219]: StringUtils.isBlank

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/NoopLoadManager.java
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/NoopLoadManager.java:151]: StringUtils.isBlank

**Review by codelipenghui — CHANGES_REQUESTED** [review]
@vineeth1995 @rdhabalia 

After reviewing all the changes, the PR introduced a new concept, `bundle affinity`, right? It's not only a new option for Admin API.

IMO, it should be 2 parts. 

- Namespace bundle affinity
- Unload the namespace bundle with affinity (or we can just split it into 2 steps, set affinity, unload bundle)

And I don't see the bundleBrokerAffinityMap will be persistent somewhere. The bundle affinity will disappear after the broker restarts?

Sorry, I have to give a `request changes` here.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:869]: If we make it async, the result could be indeterministic. If the bundle-broker mapping is not set by the time loadmanager loads the bundle, it will unload the bundle onto wrong broker.

**Comment by rdhabalia** [comment]
> And I don't see the bundleBrokerAffinityMap will be persistent somewhere. The bundle affinity will disappear after the broker restarts?

Yes, that's the purpose. we don't want persistent mapping it's useful when you want to debug by moving bundle temporarily to the specific broker. We already have namespace isolation policy for persistent affinity.

> IMO, it should be 2 parts.

No, it's connected to each other where you want to unload bundle by setting specific broker affinity for temporary testing and you don't want to keep it persistent. Here, the main idea is to keep it non-persistent for quick testing, which is the main motivation to help users in debugging.

> Sorry, I have to give a request changes here.

Let me know if you have any more concern so, we can remove the hold.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java
Comment (rdhabalia) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:869]: yes, it's dependent and we don't have to make it async here for this admin-api.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java]: addressed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java]: addressed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/NoopLoadManager.java
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/NoopLoadManager.java]: addressed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerWrapper.java
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerWrapper.java:152]: addressed

**Comment by codelipenghui** [comment]
@rdhabalia Thanks for the explanation. Clear for me now.

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:914]: ExecutionException may mean many things,
We should unwrap the cause

**Review by eolivelli — CHANGES_REQUESTED** [review]
Very good 
I generally support this work.
I left one minor comment that is worth to be addressed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:891]: We should always set a timeout, otherwise the thread may be stuck forever

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:915]: addressed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java
Comment (vineeth1995) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java]: addressed

**Comment by vineeth1995** [comment]
@eolivelli I addressed your comments.

**Comment by rdhabalia** [comment]
@eolivelli can you remove the blocker?

**Review by eolivelli — APPROVED** [review]

**Comment by eolivelli** [comment]
Lgtm

Please rebase.
We moved the website to another repository, can you please send the docs update to the website repository please? @tisonkun maybe can help

**Comment by tisonkun** [comment]
* Conflict resolved in https://github.com/apache/pulsar/pull/18663/commits/6f94be4e56f0b0b34ac95af0d0bac4dda2551bf8
* Counterpart in the site repo in https://github.com/apache/pulsar-site/pull/353

Since this is the first case, I made it a demo to be referred to later. Hopefully, we don't have new cases, but we only need to clear the backlog.

**Comment by tisonkun** [comment]
Merging...

**Pull request merged**
